libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+    }
+    if (now - threat->last_gcs_report_time > _gcs_notify_interval * 1000) {
+        GCS_MAVLINK::send_collision_all(*threat, mav_avoidance_action());
+        threat->last_gcs_report_time = now;
+    }
+
+}
+
+void AP_Avoidance::check_for_threats()
+{
+    Location my_loc;
+    if (!_ahrs.get_position(my_loc)) {
+        // if we don't know our own location we can't determine any threat level
+        return;
+    }
+
+    Vector3f my_vel;
+    if (!_ahrs.get_velocity_NED(my_vel)) {
+        // assuming our own velocity to be zero here may cause us to
+        // fly into something.  Better not to attempt to avoid in this
+        // case.
+        return;
+    }
+
+    // we always check all samples to see if they are threats since it
+    // is most likely our own position and/or velocity have changed
+    // determine the current most-serious-threat
+    _current_most_serious_threat = -1;
+    for (uint8_t i=0; i<_sample_count; i++) {
+
+        AP_Avoidance::Sample &sample = _samples[i];
+        const uint32_t sample_age = AP_HAL::millis() - sample.timestamp_ms;
+        debug("i=%d src_id=%d timestamp=%u age=%d\n", i, sample.src_id, sample.timestamp_ms, sample_age);
+
+        update_threat_level(my_loc, my_vel, sample);
+        debug("   threat-level=%d", sample.threat_level);
+
+        // ignore any really old data:
+        if (sample_age > MAX_SAMPLE_AGE_MS) {
+            continue;
+        }
+
+        bool sample_is_more_serious_threat = false;
+        do {
+            if (_current_most_serious_threat == -1) {
+                sample_is_more_serious_threat = true;
+                break;
+            }
+            if (sample.threat_level > _samples[_current_most_serious_threat].threat_level) {
+                sample_is_more_serious_threat = true;
+                break;
+            }
+            if (sample.threat_level == _samples[_current_most_serious_threat].threat_level &&
+                sample.time_to_closest_approach < _samples[_current_most_serious_threat].time_to_closest_approach) {
+                sample_is_more_serious_threat = true;
+                break;
+            }
+        } while(0);
+
+        if (sample_is_more_serious_threat) {
+            _current_most_serious_threat = i;
+        }
+    }
+    if (_current_most_serious_threat != -1) {
+        debug("Current most serious threat: %d level=%d", _current_most_serious_threat, _samples[_current_most_serious_threat].threat_level);
+    }
+}
+
+void AP_Avoidance::internal_error()
+{
+    // internal_errors++;
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    abort();
+#endif
+}
+
+AP_Avoidance::Sample *AP_Avoidance::most_serious_threat()
+{
+    if (_current_most_serious_threat < 0) {
+        // we *really_ should not have been called!
+        return nullptr;
+    }
+    return &_samples[_current_most_serious_threat];
+}
+
+
+void AP_Avoidance::update()
+{
+    if (!check_startup()) {
+        return;
+    }
+
+    get_adsb_samples();
```

Comment:
AP_Avoidance should be somewhat sensor agnostic, right? if so, we should not assume ADS-B is the only sensor. Maybe add something like if (adsb.isEnabled) { get_adsb_samples(); }. I know that screws up the flow elsewhere but this library isn't supposed to be dedicated to adsb otherwise we would just roll it into the adsb lib